src/IB/CIBStochasticMethod.cpp

```diff
@@ -0,0 +1,463 @@
+    // and translate the body to newer position.
+    const int coarsest_ln = 0;
+    const int finest_ln = d_hierarchy->getFinestLevelNumber();
+    for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+    {
+        if (!d_l_data_manager->levelContainsLagrangianData(ln)) continue;
+
+        const boost::multi_array_ref<double, 2>& X0_array =
+            *(d_l_data_manager->getLData("X0_unshifted", ln)->getLocalFormVecArray());
+        const Pointer<LMesh> mesh = d_l_data_manager->getLMesh(ln);
+        const std::vector<LNode*>& local_nodes = mesh->getLocalNodes();
+
+        // Get structures on this level.
+        const std::vector<int> structIDs = d_l_data_manager->getLagrangianStructureIDs(ln);
+        const unsigned structs_on_this_ln = (unsigned)structIDs.size();
+#if !defined(NDEBUG)
+        TBOX_ASSERT(structs_on_this_ln == d_num_rigid_parts);
+#endif
+
+        for (std::vector<LNode*>::const_iterator cit = local_nodes.begin(); cit != local_nodes.end(); ++cit)
+        {
+            const LNode* const node_idx = *cit;
+            const int lag_idx = node_idx->getLagrangianIndex();
+            const int local_idx = node_idx->getLocalPETScIndex();
+            const double* const X0 = &X0_array[local_idx][0];
+
+            int struct_handle = 0;
+            if (structs_on_this_ln > 1) struct_handle = getStructureHandle(lag_idx);
+
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                dr[d] = X0[d] - d_center_of_mass_initial[struct_handle][d];
+            }
+
+            // Rotate dr vector using the rotation matrix.
+            rotation_mat_new = d_quaternion_new[struct_handle].toRotationMatrix();
+            R_dr_new = rotation_mat_new * dr;
+
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                X_new(d) = d_center_of_mass_new[struct_handle][d] + R_dr_new(d);
+                if (!periodic_shift[d])
+                {
+                    if (X_new(d) < domain_x_lower[d] || domain_x_upper[d] < X_new(d))
+                    {
+                        reject++;
+                        pout << "REJECTED BY WALL\n";
+                    }
+                }
+            }
+        }
+        d_l_data_manager->getLData("X0_unshifted", ln)->restoreArrays();
+    }
+
+    reject = SAMRAI_MPI::sumReduction(reject);
+    if (reject)
+    {
+        d_reject = true;
+        d_num_reject++;
+    }
+
+    return;
+} // checkLagUpdate
+
+bool
+CIBStochasticMethod::getReject()
+{
+    return d_reject;
+}
+
+int
+CIBStochasticMethod::getNumReject()
+{
+    return d_num_reject;
+}
+
+void
+CIBStochasticMethod::moveLagrangianData(double delta)
+{
+    const int coarsest_ln = 0;
+    const int finest_ln = d_hierarchy->getFinestLevelNumber();
+    const double dt = delta;
+
+    // Fill the rotation matrix of structures with rotation angle 0.5*(W^n)*dt.
+    std::vector<Eigen::Matrix3d> rotation_mat(d_num_rigid_parts, Eigen::Matrix3d::Identity(3, 3));
```

Comment:
This won't work due to the custom allocation requirements with eigen. I will fix this.